docs2/site/docs/getting-started/custom-scalars.md

```diff
@@ -183,105 +211,216 @@ public class ExampleSchema : Schema
+Keep in mind that a null value also required to be handled.
 
-Implement `ValueNode<Vector3>`. Instances of `Vector3` parsed from variables will be wrapped in this type during execution.
+It is highly recommended that the type of data this method returns match the same type as
+is handled within `ParseValue`. In this case, serialization of `Vector3` is serialized as
+a string, which matches the expected data type within `ParseValue`.
 
 ```csharp
-public class Vector3Value: ValueNode<Vector3>
+// In Vector3Type
+
+public override object Serialize(object value)
 {
-    public Vector3Value(Vector3 value)
+    if (value == null)
+        return null;
+
+    if (value is Vector3 vector3)
     {
-        Value = value;
+        return $"{vector3.X},{vector3.Y},{vector3.Z}";
     }
 
-    protected override bool Equals(ValueNode<Vector3> node)
-    {
-        return Value.Equals(node.Value);
+    return ThrowSerializationError(value);
+}
+```
+
+6. Override `ScalarGraphType.ToAST` if necessary.
+
+The infrastructure converts default field values to AST representations during initialization
+in order to verify that the default values are valid within an AST tree. The default implementation
+calls `Serialize` to convert the value to its client-side equivalent, then embeds it into an
+appropriate AST node based on the data type returned from `Serialize`. It is likely you will
+only need to override this method if you are creating a custom scalar that returns enumeration
+values, or if you are returning structured data.
+
+In this example, you created a custom scalar. In summary:
+
+- Create a class for the server-side representation of the scalar
+- Implement a `ScalarGraphType` which handles variable parsing, literal parsing, and serialization
+- Register the `ScalarGraphType` with the DI container
+
+You can also choose to override `CanParseLiteral`, `CanParseValue` or `IsValidDefault` for
+enhanced performance. The default implementations call `ParseLiteral`, `ParseValue` and
+`ToAST` respectively, returning `false` if an exception is caught, or `true` otherwise.
+If you do choose to implement these methods, note that those methods must not throw an
+exception, and that they are not always called when executing a document.
+
+## Vector3 sample with combined string/structured parsing and serialization
+
+Now to extend our sample, let's assume that we want the Vector3 scalar to instead accept and
+return data in a more structured format, in addition to support the string format for variables.
+
+Here is a sample of a variable supporting a more structured format:
+
+```graphql
+mutation AddVector($vector3: Vector3!) {
+    addVector(vector3: $vector3)
+}
+
+//variables
+{
+    "vector3": {
+        "X":"23",
+        "Y":"43",
+        "Z":"66"
+    }
+}
+```
+
+And a sample of a response with a vector in a more structured format:
+
+```json
+{
+    "data": {
+        "getVector": {
```

Comment:
I'm not sure that this is the right example (although he was here before). Scalar is to determine the leaf nodes of the request/response tree. Yes, custom scalar can have some kind of formatting, but they cannot add additional levels in the tree. These are no longer scalar, but objects.

Reply:
The old 'custom scalars' page explicitly gave this sample, with a structured object being returned.

https://graphql-dotnet.github.io/docs/getting-started/custom-scalars

🤷‍♂️

I actually think it's a pretty neat sample, and demonstrates the power of custom scalars, including a need for `ToAST`.  But maybe we should add a note that the resulting schema is not spec-compliant.

Or perhaps we come up with another sample that can demonstrate a need for `ToAST`, and remove this one.

> These are no longer scalar, but object

True...  but whose fault is that?

Do we say that serialization and variables are outside the scope of graphql (as per spec), and therefor Vector3 is a scalar?  Along that logic, we look and see that ParseValue is explicitly looking for an object (dictionary) rather than a Vector3.  Why?  Because the json serialization engine cannot properly handle deserialization of Vector3.  Nor serialization for that matter (since it gets serialized as an object).  So we added a patch to ParseValue to patch a broken deserialization engine (e.g. `System.Text.Json`) that does not support deserializing Vector3 natively.

Or do we say that the graphql scalar should be written from the premise that a json serialization engine is used?  Then we can say that we should not be returning a Vector3 from Serialize because that type will be serialized as an object, which is "illegal" from graphql's POV.  (It's not a scalar anymore.)

We can even say that we "invented" a custom transport -- MyJSON -- where `{x:1,y:2,z:3}` is a Vector3 scalar.

I would generally write all of the code from the POV that a json serialization engine will be used.  Which is why I would remove native types (e.g. Guid or Uri) from being handled by ParseValue.  But in those cases at least it is a valid external representation of the value, and no harm will come from supporting it.

Anyway, back to the original question: I think we should provide a sample of a custom scalar with a reasonable premise and enough complexity so that the tooling can be demonstrated well.  It does not have to be a complex version of Vector3 of course.  But hopefully something that requires a `ToAST` implementation.

Reply:
> But maybe we should add a note that the resulting schema is not spec-compliant.

Probably. See http://spec.graphql.org/June2018/#sec-Scalars:
![изображение](https://user-images.githubusercontent.com/21261007/110234414-689f4b00-7f3b-11eb-9527-a6bb66abe484.png)


> True... but whose fault is that?

I would say ours, because we show people how to use scalar incorrectly.

> Or do we say that the graphql scalar should be written from the premise that a json serialization engine is used?

GraphQL response is **always** JSON.

> Which is why I would remove native types (e.g. Guid or Uri) from being handled by ParseValue. But in those cases at least it is a valid external representation of the value, and no harm will come from supporting it.

We should not delete them.

> But hopefully something that requires a ToAST implementation.

I already thought about JSON scalar. It will take some kind of library to implement it, so it cannot be placed in the main project. We can place that scalar in our projects for serialization. But for now just try to make an example.

Reply:
https://github.com/graphql/graphql-spec/issues/826

Reply:
> GraphQL response is always JSON.

And variables is not?

Reply:
Well, we can simply delete the second half of the sample.

Reply:
> GraphQL response is always JSON.

Not according to the specification.

http://spec.graphql.org/June2018/#sec-JSON-Serialization

> JSON is the most common serialization format for GraphQL. Though as mentioned above, GraphQL does not require a specific serialization format.

Reply:
The GraphQL spec defines the GraphQL query language, but does not make any requirements on the format of variables or result.  JSON is the most common format for variables and results to be formatted as.  I suggest limiting us to only JSON.  Then Guid and Uri and Enum are types that are not supported by JSON and hence have no need to be included within ParseValue.  And returning an object is invalid for a scalar to do.

But if we are designing the library where JSON is not a required input/output format -- for instance, perhaps the binary serializer within `System.Runtime.Serialization` was used -- then perhaps Guid and Uri and Enum are types that should be supported by ParseValue.  But also Vector3 is a completely valid scalar value, just as much as Guid or Uri.  Just like Uri, it can be serialized to a string, or can be split into parts (e.g. Scheme, Host, Port, Path, Query, etc).  There is no requirement that it MUST be serialized to a string.  It is simply a scalar that can be serialized by the binary transport in its native format.

So we can design this from a POV that only JSON transports are used, or not.

Reply:
Good news in https://github.com/graphql/graphql-spec/issues/826 It seems that it is quite possible to serialize the scalar to objects:

> the value returned by custom scalar for the result coercion (in GraphQL.js this function is called serialize) can by anything that the Programming language/environment allows. For example it can be a Money object.

> GraphQL in general doesn't not limit at all this aspect: it is an implementation detail of the GraphQL implementation.

> To phrase it a bit different: The result of an GraphQL execution is tree, where Scalars or Enums are the leaves from a type perspective. But what the actual values of these types are at runtime can't be predicted or described by the spec.

I suggest you copy these phrases as is into documentation (with some changes). You can do this at the very beginning of a more complex second example to explain the motivation.

Please also add somewhere in text links to https://www.graphql-scalars.com/scalars-guide and https://www.graphql.de/blog/scalars-in-depth/ .